src/framework/MOM_murmur_hash.F90

```diff
@@ -0,0 +1,251 @@
+!> MurmurHash is a non-cryptographic hash function developed by Austin Appleby.
+!!
+!! This module provides an implementation of the 32-bit MurmurHash3 algorithm.
+!! It is used in MOM6 to generate unique hashes of field arrays.  The hash is
+!! sensitive to order of elements and can detect changes that would otherwise
+!! be missed by the mean/min/max/bitcount tests.
+!!
+!! Sensitivity to order means that it must be used with care for tests such as
+!! processor layout.
+!!
+!! This implementation assumes data sizes of either 32 or 64 bits.  It cannot
+!! be used for smaller types such as strings.
+!!
+!! https://github.com/aappleby/smhasher
+module MOM_murmur_hash
+
+use, intrinsic :: iso_fortran_env, only : int32, int64, real32, real64
+
+implicit none ; private
+
+public :: murmur_hash
+
+!> Return the murmur3 hash of an array.
+interface murmur_hash
+  procedure murmurhash3_i32
+  procedure murmurhash3_i64
+  procedure murmurhash3_r32
+  procedure murmurhash3_r32_1d
+  procedure murmurhash3_r32_2d
+  procedure murmurhash3_r32_3d
+  procedure murmurhash3_r32_4d
+  procedure murmurhash3_r64
+  procedure murmurhash3_r64_1d
+  procedure murmurhash3_r64_2d
+  procedure murmurhash3_r64_3d
+  procedure murmurhash3_r64_4d
+end interface murmur_hash
+
+contains
+
+!> Return the murmur3 hash for a 32-bit integer array.
+function murmurhash3_i32(key, seed) result(hash)
+  integer(int32), intent(in) :: key(:)
+    !< Input array
+  integer(int32), intent(in), optional :: seed
+    !< Hash seed
+  integer(int32) :: hash
+    !< Murmur hash of array
+
+  integer(int32), parameter :: c1 = int(z'cc9e2d51', kind=int32)
+  integer(int32), parameter :: c2 = int(z'1b873593', kind=int32)
+  integer(int32), parameter :: c3 = int(z'e6546b64', kind=int32)
+
+  integer(int32), parameter :: c4 = int(z'85ebca6b', kind=int32)
+  integer(int32), parameter :: c5 = int(z'c2b2ae35', kind=int32)
+
+  integer :: i
+  integer(int32) :: k
+
+  hash = 0
+  if (present(seed)) hash = seed
+
+  do i = 1, size(key)
+    k = key(i)
+    k = k * c1
+    k = ishftc(k, 15)
+    k = k * c2
+
+    hash = ieor(hash, k)
+    hash = ishftc(hash, 13)
+    hash = 5 * hash + c3
+  enddo
+
+  ! NOTE: This is the point where the algorithm would handle trailing bytes.
+  ! Since our arrays are comprised of 4 or 8 byte elements, we skip this part.
+
+  hash = ieor(hash, 4*size(key))
+
+  hash = ieor(hash, shiftr(hash, 16))
```

Comment:
This PR is not compiling with the PGI / NVIDIA 2023 compiler, which apparently does not support the Fortran 2008 standard `shiftr()` and `shiftl()` functions.  Could this function be implemented without using `shiftr()`?

Reply:
If `shifta` is available then this seems to pass my unit tests:
```
s = shifta(i, 1)
s = ibclr(s, 31)
s = shifta(i, shift-1)
```
although if there's no `shiftl` then I can't imagine it would have `shifta`.

This is basically the same idea:
```
  s = s / 2
  s = ibclr(s, 31)
  s = s / 2**(shift-1)
```
These are kind of ridiculous since none of them use the assembly instruction.  But maybe I could make a function for conditional use.

Reply:
The standard has an even simpler suggestion
```
s = ishft(i, -shift)
```
That's assuming Nvfortran can do `ishft`...

Reply:
`ishft()` seems to be in 23.9, let's go with that.